src/toxcore/dht/ping_checker.rs

```diff
@@ -0,0 +1,213 @@
+/*
+    Copyright © 2017 Zetok Zalbavar <zexavexxe@gmail.com>
+    Copyright © 2018 Namsoo CHO <nscho66@gmail.com>
+
+    This file is part of Tox.
+
+    Tox is libre software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    Tox is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with Tox.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+
+/*!
+Manage ping_id.
+Generate ping_id on request packet, check ping_id on response packet.
+*/
+
+use std::io::{ErrorKind, Error};
+use std::collections::HashMap;
+use std::time::{Duration, Instant};
+
+use toxcore::crypto_core::*;
+use toxcore::dht::packed_node::*;
+use toxcore::dht::server::*;
+
+/// Entry type of PingChecker
+#[derive(Clone, Debug, PartialEq)]
+struct PingCheckerEntry {
+    data: Vec<PackedNode>,
+    time: Instant,
+    ping_id: u64,
+}
+
+impl PingCheckerEntry {
+    /// Check if it is timedout
+    pub fn is_timeout(&self, timeout: Duration) -> bool {
+        self.time.elapsed() > timeout
+    }
+}
+
+/// Generate ping_id, check if ping_id is correct or timedout
+#[derive(Clone, Debug, PartialEq)]
+pub struct PingChecker {
+    entries: HashMap<u64, PingCheckerEntry>,
+    last_added: u64,
+    last_deleted: u64,
+    total_size: u64,
+}
+
+impl PingChecker {
+    /// Create new PingChecker object
+    pub fn new(total_size: u64) -> PingChecker {
+        assert!(total_size > 0);
+
+        PingChecker {
+            entries: HashMap::new(),
+            last_added: 0,
+            last_deleted: 0,
+            total_size,
+        }
+    }
+
+    // Clear timed out entries
+    fn clear_timedout(&mut self, timeout: Duration) {
+        while self.last_deleted != self.last_added {
+            if let Some(ping_entry) = self.entries.get(&self.last_deleted) {
+                if !ping_entry.is_timeout(timeout) {
+                    break;
+                }
+            }
+
+            self.entries.remove(&self.last_deleted);
+            self.last_deleted = (self.last_deleted + 1) % self.total_size;
+        }
+    }
+
+    /// Add a data to the PingChecker and return a new ping_id.
+    pub fn add(&mut self, data: Vec<PackedNode>) -> u64 {
```

Comment:
not `data`, but `nodes`.

it is not obvious why there is a possibility to add multiple nodes